pykrylov/linop/lsr1.py

```diff
@@ -349,21 +349,27 @@ def qn_matvec(self, v):
         for i in range(npairs):
             k = (self.insert + i) % npairs
             if ys[k] is not None:
+                # Form all a[] and ad[] vectors for the current step
                 a[:, k] = y[:, k] - s[:, k] / self.gamma
                 ad[:, k] = yd[:, k] - s[:, k] / self.gamma
                 for j in range(i):
                     l = (self.insert + j) % npairs
                     if ys[l] is not None:
-                        alTs = np.dot(a[:, l], s[:, k])
-                        adlTs = np.dot(ad[:, l], s[:, k])
-                        update = -alTs / aTs[l] * ad[:, l] - adlTs / aTs[l] * \
-                            a[:, l] + adTs[l] / aTs[l] * alTs * a[:, l]
-                        a[:, k] += update.copy()
-                        ad[:, k] += update.copy()
+                        aTsk = np.dot(a[:, l], s[:, k])
+                        adTsk = np.dot(ad[:, l], s[:, k])
+                        aTsl = np.dot(a[:, l], s[:, l])
+                        adTsl = np.dot(ad[:, l], s[:, l])
+                        update = (aTsk / aTsl) * ad[:, l] + (adTsk / aTsl) * a[:, l] - \
+                            (aTsk * adTsl / aTsl**2) * a[:, l]
+                        a[:, k] -= update.copy()
+                        ad[:, k] -= update.copy()
```

Comment:
Why these changes? It's hard to tell whether they actually modify the behavior or are just a rewrite.